In [1]:
import pandas as pd
melb_df = pd.read_csv('data/citibike-tripdata.csv', sep=',')
#data_df = pd.read_csv('https://raw.githubusercontent.com/justmarkham/pandas-videos/master/data/ufo.csv')
# melb_df.head()
# melb_df.shape # (300000, 14)
# melb_df.info()
melb_df.describe(include=['object'])

starttime                  stoptime  \
count                     300000                    300000   
unique                    299836                    299813   
top     2018-09-02 13:59:41.7770  2018-09-04 17:51:07.9110   
freq                           2                         2   

                       start station name       end station name    usertype  
count                              299831                 299831      300000  
unique                                759                    765           2  
top     Grand Army Plaza & Central Park S  West St & Chambers St  Subscriber  
freq                                 1928                   1963      232202

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   starttime                300000 non-null  object 
 1   stoptime                 300000 non-null  object 
 2   start station id         299831 non-null  float64
 3   start station name       299831 non-null  object 
 4   start station latitude   300000 non-null  float64
 5   start station longitude  300000 non-null  float64
 6   end station id           299831 non-null  float64
 7   end station name         299831 non-null  object 
 8   end station latitude     300000 non-null  float64
 9   end station longitude    300000 non-null  float64
 10  bikeid                   300000 non-null  int64  
 11  usertype                 300000 non-null  object 
 12  birth year               300000 non-null  int64  
 13  gender                   300000 non-null  int64  
dtypes: float64(6), int64(3), object(5)
memory usage: 32.0+ MB

starttime — время начала поездки (дата, время);
stoptime — время окончания поездки (дата, время);
start station id — идентификатор стартовой стоянки;
start station name — название стартовой стоянки;
start station latitude, start station longitude — географическая широта и долгота стартовой стоянки;
end station id — идентификатор конечной стоянки;
end station name — название конечной стоянки;
end station latitude, end station longitude — географическая широта и долгота конечной стоянки;
bikeid — идентификатор велосипеда;
usertype — тип пользователя (Customer — клиент с подпиской на 24 часа или на три дня, Subscriber — подписчик с годовой арендой велосипеда);
birth year — год рождения клиента;
gender — пол клиента (0 — неизвестный, 1 — мужчина, 2 — женщина).

6.1

In [2]:
melb_df['start station id'].isnull().sum ()

169

6.2

In [3]:
# melb_df['starttime'].dtype
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   starttime                300000 non-null  object 
 1   stoptime                 300000 non-null  object 
 2   start station id         299831 non-null  float64
 3   start station name       299831 non-null  object 
 4   start station latitude   300000 non-null  float64
 5   start station longitude  300000 non-null  float64
 6   end station id           299831 non-null  float64
 7   end station name         299831 non-null  object 
 8   end station latitude     300000 non-null  float64
 9   end station longitude    300000 non-null  float64
 10  bikeid                   300000 non-null  int64  
 11  usertype                 300000 non-null  object 
 12  birth year               300000 non-null  int64  
 13  gender                   300000 non-null  int64  
dtypes: f

6.3

In [4]:
# melb_df['start station id'].value_counts()
# melb_df['start station id'].describe()
# melb_df['start station id'].value_counts()
melb_df['start station id'].mode()


0    281.0
Name: start station id, dtype: float64

6.4

In [5]:
melb_df['bikeid'].mode()

0    33887
Name: bikeid, dtype: int64

6.5

In [6]:
# melb_df[(melb_df['usertype']=='Subscriber') | (melb_df['usertype']=='Customer')]['usertype']
round(melb_df['usertype'].value_counts(normalize=True),2)

usertype
Subscriber    0.77
Customer      0.23
Name: proportion, dtype: float64

In [7]:
mode_usertype = melb_df['usertype'].mode()[0]
count_mode_user = melb_df[melb_df['usertype'] == mode_usertype].shape[0]
print(round(count_mode_user / melb_df.shape[0], 2))

0.77


6.6

In [8]:
melb_df['gender'].value_counts().max()

183582

In [9]:
male_count = melb_df[melb_df['gender'] == 1].shape[0]
female_count = melb_df[melb_df['gender'] == 0].shape[0]
print(max([male_count, female_count]))

183582


6.7

pandas.to_datetime

In [10]:
print('Число уникальных стартовых и конечных стоянок')
print(melb_df['start station id'].nunique())
print(melb_df['end station id'].nunique())

# # melb_df.describe().loc[:, ['start station name','end station name']]

# # найти год старта - год раждения min
melb_df['starttime'] = pd.to_datetime(melb_df['starttime'], yearfirst=True)
melb_df['stoptime']  = pd.to_datetime(melb_df['stoptime'],errors='coerce',  yearfirst=True)
min_y=melb_df['starttime'].dt.year.min()

# display(melb_df.loc[:,['starttime','stoptime']])
print('\n','минимальный возвраст участника', min_y-melb_df['birth year'].max())
print('\n','Самой непопулярной стартовой стоянкой')
print(melb_df['start station name'].value_counts()) # Eastern Pkwy & Washington Ave 6
print('\n','На какой стоянки Наибольшее количество поездок завершается')
print(melb_df['end station name'].value_counts().head()) # West St & Chambers St  1963

Число уникальных стартовых и конечных стоянок
759
765

 минимальный возвраст участника 16

 Самой непопулярной стартовой стоянкой
start station name
Grand Army Plaza & Central Park S    1928
Central Park S & 6 Ave               1909
Pershing Square North                1873
12 Ave & W 40 St                     1845
West St & Chambers St                1835
                                     ... 
NYCBS Depot - GOW                      15
Franklin Ave & Empire Blvd             11
Railroad Ave & Kay Ave                  9
47 Ave & Skillman Ave                   7
Eastern Pkwy & Washington Ave           6
Name: count, Length: 759, dtype: int64

 На какой стоянки Наибольшее количество поездок завершается
end station name
West St & Chambers St     1963
12 Ave & W 40 St          1909
Pershing Square North     1850
Central Park S & 6 Ave    1828
E 17 St & Broadway        1804
Name: count, dtype: int64


6.8

In [11]:
# melb_df.head()
print(melb_df['start station id'].nunique())
print(melb_df['start station name'].nunique())
print(melb_df['end station id'].nunique())
print(melb_df['end station name'].nunique())
print('Сколько столбцов',melb_df.shape[1])
melb_df.drop(['start station name','end station name'],axis=1,inplace=True)
print('Сколько столбцов осталось',melb_df.shape[1])

759
759
765
765
Сколько столбцов 14
Сколько столбцов осталось 12


6.9

In [12]:
melb_df['age']=min_y-melb_df['birth year']
melb_df.drop(['birth year'],axis=1,inplace=True)

In [13]:
age60=melb_df[melb_df['age']>60]['age'].count()
print('клиенты старше 60:',age60)

клиенты старше 60: 11837


6.10

In [14]:
# data_time=lambda x: (x.dt.components['minutes']+x.dt.components['hours']*60)
melb_df['trip duration']=melb_df['stoptime']-melb_df['starttime']
print('длилась поездка под индексом 3=',melb_df.loc[3,'trip duration'])

длилась поездка под индексом 3= 0 days 00:07:16.837000


In [15]:
melb_df['starttime'] = pd.to_datetime(melb_df['starttime'])
melb_df['stoptime'] = pd.to_datetime(melb_df['stoptime'])
melb_df['trip duration'] = (melb_df['stoptime'] - melb_df['starttime'])
melb_df.loc[3, 'trip duration']

Timedelta('0 days 00:07:16.837000')

6.11
признак-мигалку

In [16]:
# weekend=melb_df['starttime'].dt.weekday
# melb_df['weekend']=weekend.apply(lambda x: 1 if x>4 else 0)
melb_df['weekend']=melb_df['starttime'].dt.weekday.apply(lambda x: 1 if x>4 else 0)
# melb_df['weekend'].value_counts()
print('сколько поездок начиналось в выходные',melb_df['weekend'].sum())
# a=3
# x=1 if a>4 else 0
# x
# melb_df['StreetType'] = street_types.apply(lambda x: x if x in popular_stypes else 'other')
# apply(lambda x: 1 if x>4 else 0)


сколько поездок начиналось в выходные 115135


6.12

In [19]:
# def hour_dt(x):
#     if x<7:          return 'night'
#     if x>6 and x<13: return 'morning'
#     if x>12 and x<19: return 'day'
#     if x>18 and x<23: return 'evening'

# melb_df['time_of_day']=melb_df['starttime'].dt.hour.apply(hour_dt)

y=['night','morning','day','evening','night']

# melb_df['time_of_day']=melb_df['starttime'].dt.hour.apply(lambda x:y[(x-1)//6] if x>0 else y[0])

melb_df['time_of_day']=melb_df['starttime'].dt.hour.apply(lambda x:y[(x-1)//6])

# day_nig = melb_df[((melb_df['time_of_day']=='night') | (melb_df['time_of_day']=='day'))]['time_of_day']
print('Во сколько раз количество поездок, совершённых днём, больше, чем количество поездок, совёршенных ночью')
print(round(melb_df['time_of_day'].value_counts()['day']/melb_df['time_of_day'].value_counts()['night']))
melb_df['time_of_day'].value_counts()

Во сколько раз количество поездок, совершённых днём, больше, чем количество поездок, совёршенных ночью
9


time_of_day
day        143012
morning     95530
evening     46373
night       15085
Name: count, dtype: int64

In [18]:
for i in range(24):
    print(i,'=',(i-1)//6 if i>0 else 0)

0 = 0
1 = 0
2 = 0
3 = 0
4 = 0
5 = 0
6 = 0
7 = 1
8 = 1
9 = 1
10 = 1
11 = 1
12 = 1
13 = 2
14 = 2
15 = 2
16 = 2
17 = 2
18 = 2
19 = 3
20 = 3
21 = 3
22 = 3
23 = 3


starttime — время начала поездки (дата, время);
stoptime — время окончания поездки (дата, время);
start station id — идентификатор стартовой стоянки;
start station name — название стартовой стоянки;
start station latitude, start station longitude — географическая широта и долгота стартовой стоянки;
end station id — идентификатор конечной стоянки;
end station name — название конечной стоянки;
end station latitude, end station longitude — географическая широта и долгота конечной стоянки;
bikeid — идентификатор велосипеда;
usertype — тип пользователя (Customer — клиент с подпиской на 24 часа или на три дня, Subscriber — подписчик с годовой арендой велосипеда);
birth year — год рождения клиента;
gender — пол клиента (0 — неизвестный, 1 — мужчина, 2 — женщина).